sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -1,73 +1,166 @@
-# Author: Andrew nystrom <awnystrom@gmail.com>
+# Authors: Andrew nystrom <awnystrom@gmail.com>
+#          Meekail Zain <zainmeekail@gmail.com>
+from libc.limits cimport INT_MAX, LONG_MAX
+from libc.math cimport sqrt, pow
+cimport numpy as cnp
+cnp.import_array()
 
 from scipy.sparse import csr_matrix
-cimport numpy as cnp
 import numpy as np
 
-cnp.import_array()
+ctypedef cnp.int8_t FLAG_t
+
+# We use the following verbatim block to determine whether the current
+# platform's compiler supports 128-bit integer values intrinsically.
+# This should work for GCC and CLANG, but doesn't for MSVC. We prefer to use
+# 128-bit integers when possible because the intermediate calculations have a
+# non-trivial risk of overflow. It is, however, very unlikely to come up on an
+# average use case, hence 64-bit integers are "good enough".
+cdef extern from *:
+    """
+    #ifdef __SIZEOF_INT128__
+        typedef __int128 LARGEST_INT_t;
+    #elif __clang__
+        typedef _BitInt(128) LARGEST_INT_t;
+    #else
+        typedef long long LARGEST_INT_t;
+    #endif
+    """
+    ctypedef long long LARGEST_INT_t
```

Comment:
I think that we might want to add a test to check for the size of `LARGEST_INT_t`.

What do you think?

Reply:
I agree but I am not sure how to set the expectations.

Maybe we can write a testing support Cython helper such as:

```cython
cpdef LARGEST_INT_t _increment_large_integer(LARGEST_INT_t value):
    return value + 1
```

and then call this helper in a test such as:

```python
def test_largest_int():
    if sys.platform == "win32":
        # On Windows, scikit-learn is typically compiled with MSVC that
        # does not support int128 arithmetic (at the time of writing):
        # https://stackoverflow.com/a/6761962/163740
        # Therefore we expect `_increment_large_integer` to overflow in the
        # following case:
        assert (
            _increment_large_integer(np.iinfo(np.int64).max) == np.iinfo(np.int64).min
        )
    else:
        # An all other supported platforms, C/C++ compilers used to build
        # scikit-learn are expected to support int128 operations:
        incremented = _increment_large_integer(np.iinfo(np.int64).max)
        assert incremented != np.iinfo(np.int64).min
        assert incremented == np.iinfo(np.int64).max + 1

```

Reply:
Would it be sufficient just to check the `sizeof` the type is either `8` or `16` based on expectation?

Reply:
Indeed.

Reply:
I have added `test_sizeof_LARGEST_INT_t()` to address this